src/Interpreters/TreeOptimizer.cpp

```diff
@@ -153,6 +153,17 @@ void optimizeGroupBy(ASTSelectQuery * select_query, ContextPtr context)
                     continue;
                 }
             }
+            /// don't optimise functions that shadow any of it's arguments:
+            /// https://github.com/ClickHouse/ClickHouse/issues/37032
+            if (std::any_of(
+                    function->arguments->children.begin(),
+                    function->arguments->children.end(),
+                    [&function](const ASTPtr & arg) { return function->alias == arg->getAliasOrColumnName(); }))
```

Comment:
This will not work for functions inside functions, I guess you need to write a visitor here:

```sql
SELECT toString(dummy+0) as dummy FROM remote('127.{1,1}', 'system.one') GROUP BY dummy;
Code: 43. DB::Exception: Illegal types String and UInt8 of arguments of function plus: While processing (toString(dummy + 0) AS dummy) + 0. (ILLEGAL_TYPE_OF_ARGUMENT)
```

Reply:
Thanks for the catch @azat . Submitted a fix, please take a look.